File: python/src/solana_keychain/crossmint/derive.py
"""Crossmint delegated-signer key derivation."""

try:
    import base58
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.hkdf import HKDF
except ImportError as error:  # pragma: no cover
    raise ImportError(
        "solana_keychain.crossmint requires the crossmint extra: "
        "pip install 'solana-keychain[crossmint]'"
    ) from error

from solders.keypair import Keypair

from solana_keychain.core.errors import SignerError, SignerErrorCode

SIGNER_SECRET_HEX_LENGTH = 64


def parse_api_key(api_key: str) -> tuple[str, str]:
    """Extract ``(project_id, environment)`` from a Crossmint API key of the form
    ``{ck|sk}_{environment}_{base58data}`` where the base58 data decodes to UTF-8
    ``projectId:nacl_signature``."""
    parts = api_key.split("_", 2)
    if len(parts) != 3:
        raise SignerError(SignerErrorCode.CONFIG_ERROR, "Invalid API key format")
    environment, base58_data = parts[1], parts[2]
    try:
        decoded = base58.b58decode(base58_data).decode("utf-8")
    except (ValueError, UnicodeDecodeError):
        raise SignerError(SignerErrorCode.CONFIG_ERROR, "Failed to decode API key data") from None
    project_id = decoded.split(":", 1)[0]
    if not project_id:
        raise SignerError(SignerErrorCode.CONFIG_ERROR, "Could not extract projectId from API key")
    return project_id, environment


def derive_signing_key(secret: str, api_key: str) -> Keypair:
    """Derive the server delegated-signer Ed25519 keypair from an ``xmsk1_``-prefixed
    64-hex-char secret via HKDF-SHA256 (salt ``crossmint``, info
    ``{projectId}:{environment}:solana-ed25519``)."""
    project_id, environment = parse_api_key(api_key)

    raw_secret = secret.removeprefix("xmsk1_")
    if len(raw_secret) != SIGNER_SECRET_HEX_LENGTH:
        raise SignerError(
            SignerErrorCode.CONFIG_ERROR,
            f"signer_secret must be a {SIGNER_SECRET_HEX_LENGTH}-char hex string "
            f"(got {len(raw_secret)})",
        )
    try:
        ikm = bytes.fromhex(raw_secret)
    except ValueError:
        raise SignerError(SignerErrorCode.CONFIG_ERROR, "signer_secret is not valid hex") from None

    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=b"crossmint",
        info=f"{project_id}:{environment}:solana-ed25519".encode(),
    )
    return Keypair.from_seed(hkdf.derive(ikm))

File: go/signers/crossmint/derive.go
package crossmint

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"strings"
	"unicode/utf8"

	"golang.org/x/crypto/hkdf"

	"github.com/solana-foundation/solana-go/v2/base58"

	"github.com/solana-foundation/solana-keychain/go/core/v2"
)

// signerSecretPrefix is the expected prefix of a Crossmint server signer secret.
const signerSecretPrefix = "xmsk1_"

// hkdfSalt is the fixed HKDF salt used by Crossmint server-signer derivation.
const hkdfSalt = "crossmint"

// deriveSigningKey derives the Ed25519 approval key from a server signer
// secret: the secret's 64 hex chars (after an optional "xmsk1_" prefix) are
// the HKDF-SHA256 IKM, the salt is "crossmint", the info string is
// "{projectId}:{environment}:solana-ed25519" (both extracted from the API
// key), and the 32-byte output is the Ed25519 seed.
func deriveSigningKey(secret, apiKey string) (ed25519.PrivateKey, error) {
	projectID, environment, err := parseAPIKey(apiKey)
	if err != nil {
		return nil, err
	}

	rawSecret := strings.TrimPrefix(secret, signerSecretPrefix)
	if len(rawSecret) != 64 {
		return nil, core.NewSignerError(core.CodeConfigError,
			fmt.Sprintf("signer_secret must be a 64-char hex string (got %d)", len(rawSecret)))
	}
	ikm, err := hex.DecodeString(rawSecret)
	if err != nil {
		return nil, core.WrapSignerError(core.CodeConfigError, "signer_secret is not valid hex", err)
	}

	info := projectID + ":" + environment + ":solana-ed25519"
	reader := hkdf.New(sha256.New, ikm, []byte(hkdfSalt), []byte(info))
	seed := make([]byte, ed25519.SeedSize)
	if _, err := io.ReadFull(reader, seed); err != nil {
		return nil, core.WrapSignerError(core.CodeConfigError, "HKDF expand failed", err)
	}
	return ed25519.NewKeyFromSeed(seed), nil
}

// parseAPIKey extracts the projectId and environment from a Crossmint API key.
// Format: {ck|sk}_{environment}_{base58data}, where the base58 data decodes to
// the UTF-8 string "projectId:nacl_signature".
func parseAPIKey(apiKey string) (projectID, environment string, err error) {
	parts := strings.SplitN(apiKey, "_", 3)
	if len(parts) < 3 {
		return "", "", core.NewSignerError(core.CodeConfigError, "invalid API key format")
	}
	environment = parts[1]

	decoded, err := base58.Decode(parts[2])
	if err != nil {
		return "", "", core.WrapSignerError(core.CodeConfigError, "failed to decode API key data", err)
	}
	if !utf8.Valid(decoded) {
		return "", "", core.NewSignerError(core.CodeConfigError, "API key data is not valid UTF-8")
	}
	projectID = strings.SplitN(string(decoded), ":", 2)[0]
	return projectID, environment, nil
}
